crypto/bio/b_sock2.c

```diff
@@ -115,6 +115,55 @@ int BIO_connect(int sock, const BIO_ADDR *addr, int options)
     return 1;
 }
 
+/*-
+ * BIO_bind - bind socket to address
+ * @sock: the socket to set
+ * @addr: local address to bind to
+ * @options: BIO socket options
+ *
+ * Binds to the address using the given socket and options.
+ *
+ * Options can be a combination of the following:
+ * - BIO_SOCK_REUSEADDR: Try to reuse the address and port combination
+ *   for a recently closed port.
+ *
+ * When restarting the program it could be that the port is still in use.  If
+ * you set to BIO_SOCK_REUSEADDR option it will try to reuse the port anyway.
+ * It's recommended that you use this.
+ */
+int BIO_bind(int sock, const BIO_ADDR *addr, int options)
+{
+    int on = 1;
+
+    if (sock == -1) {
+        BIOerr(BIO_F_BIO_BIND, BIO_R_INVALID_SOCKET);
+        return 0;
+    }
+
+# ifndef OPENSSL_SYS_WINDOWS
+    /*
+     * SO_REUSEADDR has different behavior on Windows than on
+     * other operating systems, don't set it there.
+     */
+    if (options & BIO_SOCK_REUSEADDR) {
+        if (setsockopt(sock, SOL_SOCKET, SO_REUSEADDR,
+                       (const void *)&on, sizeof(on)) != 0) {
+            SYSerr(SYS_F_SETSOCKOPT, get_last_socket_error());
+            BIOerr(BIO_F_BIO_BIND, BIO_R_UNABLE_TO_REUSEADDR);
+            return 0;
+        }
+    }
+# endif
```

Comment:
I wonder what's rationale behind moving this from listen? It should be noted that it kind of goes against documentation, which says that BIO_SOCK_REUSEADDR is available for BIO_listen. Well, this inevitably calls for question about documentation.

Reply:
It needs to be done here because we may need to bind multiple sockets to the same calling address -- for example some ftp implementations insist that the data connection always comes from port 20.  (Yes, it's mad, but it is the second oldest protocol on the internet).

BIO_SOCK_REUSEADDR is still available for BIO_listen because BIO_listen calls BIO_bind.

Documentation is, of course, needed.

Reply:
Documentation added.

Reply:
I suppose what tipped me off was that originally it started as something about one of interfaces' *addresses*, while REUSEADDR is actually rather about *port*... In the context one can wonder if it would be appropriate if caller, apps/s_socket.c, passed REUSE address only if user asked for specific port...

Reply:
On side note one can probably argue that for **client** sockets it would be appropriate to use REUSEADDR on Windows. But we can't tell the difference, so we let it to application...